lib/src/subscription.dart

```diff
@@ -0,0 +1,155 @@
+}
+
+enum SubscriptionSetState {
+  uncommitted,
+  pending,
+  bootstrapping,
+  complete,
+  error,
+  superseded,
+}
+
+abstract class SubscriptionSet with IterableMixin<Subscription> {
+  Realm _realm;
+  SubscriptionSetHandle _handle;
+
+  SubscriptionSet._(this._realm, this._handle);
+
+  Subscription? find<T extends RealmObject>(RealmResults<T> query) {
+    return realmCore.findSubscriptionByQuery(this, query).convert(Subscription._);
+  }
+
+  Subscription? findByName(String name) {
+    return realmCore.findSubscriptionByName(this, name).convert(Subscription._);
+  }
+
+  Future<SubscriptionSetState> waitForStateChange(SubscriptionSetState state) async {
+    return SubscriptionSetState.values[await realmCore.waitForSubscriptionSetStateChange(this, state)];
+  }
+
+  @override
+  int get length => realmCore.getSubscriptionSetSize(this);
+
+  @override
+  Subscription elementAt(int index) {
+    return Subscription._(realmCore.subscriptionAt(this, index));
+  }
+
+  Subscription operator [](int index) => elementAt(index);
+
+  @override
+  _SubscriptionIterator get iterator => _SubscriptionIterator._(this);
+
+  void update(void Function(MutableSubscriptionSet mutableSubscriptions) action);
+
+  int get version => realmCore.subscriptionSetVersion(this);
+
+  SubscriptionSetState get state => SubscriptionSetState.values[realmCore.subscriptionSetState(this)];
+}
+
+extension SubscriptionSetInternal on SubscriptionSet {
+  Realm get realm => _realm;
+  SubscriptionSetHandle get handle => _handle;
+
+  static SubscriptionSet create(Realm realm, SubscriptionSetHandle handle) => MutableSubscriptionSet._(realm, handle);
+}
+
+class MutableSubscriptionSet extends SubscriptionSet {
+  MutableSubscriptionSetHandle? _mutableHandle;
+
+  MutableSubscriptionSet._(Realm realm, SubscriptionSetHandle handle) : super._(realm, handle);
+
+  @override
+  void update(void Function(MutableSubscriptionSet mutableSubscriptions) action) {
+    assert(_mutableHandle == null);
+    var commit = false;
+    try {
+      _mutableHandle = realmCore.makeSubscriptionSetMutable(this);
+      action(this);
+      commit = true;
+    } finally {
+      if (commit) {
+        _handle = realmCore.subscriptionSetCommit(this);
+      }
+      // _mutableHandle.release(); // TODO: Release early
+      _mutableHandle = null;
+    }
+  }
+
+  bool addOrUpdate<T extends RealmObject>(RealmResults<T> query, {String? name}) {
+    assert(_mutableHandle != null);
+    return realmCore.insertOrAssignSubscription(this, query, name);
+  }
+
+  void remove<T extends RealmObject>(RealmResults<T> query) {
+    assert(_mutableHandle != null);
+    return realmCore.eraseSubscriptionByQuery(this, query);
+  }
+
+  void removeByName(String name) {
+    assert(_mutableHandle != null);
+    return realmCore.eraseSubscriptionByName(this, name);
+  }
+
+  void removeAll() {
+    assert(_mutableHandle != null);
+    return realmCore.clearSubscriptionSet(this);
+  }
```

Comment:
Regarding return values. They will all either succeed or throw.

Reply:
Isn't that what `OwnableHandle` do? :-) ..

Anyway, I would like to avoid `ByQuery` to maintain symmetry with `addOrUpdate`.

Also, a subscription can know what set it belongs to, so you avoid matching them up correctly, ie. `Subscription.remove()` doesn't need an argument.

We can totally have the method on the `SubscriptionSet` but I would like to keep the existing names, so what about `removeSubscription`?

Reply:
Changed